File: gfg/2024_7_july/python/12_root_to_leaf_path_sum.py
# CODE

class Node:
    def __init__(self,val):
        self.data = val
        self.left = None
        self.right = None

class Solution:
    def hasPathSum(self, root, target):
        if root is None:
            return False
        if root.left is None and root.right is None:
            return target == root.data
        rem_sum = target - root.data
        return (self.hasPathSum(root.left, rem_sum) or 
                self.hasPathSum(root.right, rem_sum))

File: gfg/2024_7_july/c++/12_root_to_leaf_path_sum.cpp
// TIME COMPLEXITY
// O(n)

// SPACE COMPLEXITY
// O(n)

// CODE

class Solution {
  public:
    bool solve(Node *root, int target){
        if(root == NULL) return false;
        
        // leaf node compare with target
        if(root->left == NULL && root->right == NULL) return (root->data == target);
        
        // if not leaf updarte target
        int tar = target - root->data;
        
        bool left = solve(root->left, tar);
        bool right = solve(root->right, tar);
        
        
        if(left) return left;
        return right;
    }
    bool hasPathSum(Node *root, int target) {
        // Your code here
        return solve(root,target);
        
    }
};